File: fsharp/magic-square/test/puzzle-test.fs
module puzzle_test

open Expecto
open magic_square.puzzle

let sum_rows (m : List<float list>) : float =
    m
    |> List.map (fun x -> x |> List.reduce (+))
    |> List.sum

let sum_cols (m : List<float list>) : float =
    [
        m |> List.map (fun x -> x |> List.head)
        m |> List.map (fun x -> x |> List.item 1)
        m |> List.map (fun x -> x |> List.last)
    ]
    |> sum_rows

let sum_diagonals (m : List<float list>) : float =
    ((m |> List.item 0) |> List.item 0)
    + ((m |> List.item 1) |> List.item 1)
    + ((m |> List.item 2) |> List.item 2)
    + ((m |> List.item 2) |> List.item 0)
    + ((m |> List.item 1) |> List.item 1)
    + ((m |> List.item 0) |> List.item 2)

[<Tests>]
let tests =
    testList "puzzle" [
        testCase "all the rows, columns, and diagonal add to the same number" <| fun _ ->
            let rows = values |> magic_square |> sum_rows
            let cols = values |> magic_square |> sum_cols
            let diag = values |> magic_square |> sum_diagonals
            Expect.isTrue (rows = cols) ""
            Expect.isTrue (rows = diag) ""
            Expect.isTrue (cols = diag) ""
    ]

[<EntryPoint>]
let main argv =
    runTestsInAssembly defaultConfig argv

File: csharp/magic-square/test/puzzle-test.cs
namespace magic_square;

using Xunit;
using System.Linq;
using System.Collections.Generic;
using static magic_square.puzzle;

public class puzzle_test {

    private double sum_rows(List<List<double>> m) =>
        m
        .Select(x => x.Sum())
        .Sum();

    private double sum_cols(List<List<double>> m) =>
        sum_rows (
            new List<List<double>>() {
                new List<double>(m.Select(x => x.First())),
                new List<double>(m.Select(x => x.ElementAt(1))),
                new List<double>(m.Select(x => x.Last()))
            }
        );

    private double sum_diagonals(List<List<double>> m) =>
        (m.ElementAt(0).ElementAt(0))
        + (m.ElementAt(1).ElementAt(1))
        + (m.ElementAt(2).ElementAt(2))
        + (m.ElementAt(2).ElementAt(0))
        + (m.ElementAt(1).ElementAt(1))
        + (m.ElementAt(0).ElementAt(2));

    [Fact]
    public void all_rows_cols_diag_add_to_same_number() {
        double rows = sum_rows(magic_square(values()));
        double cols = sum_cols(magic_square(values()));
        double diag = sum_diagonals(magic_square(values()));
        Assert.True(rows == cols);
        Assert.True(rows == diag);
        Assert.True(cols == diag);
    }
}
